stan/math/prim/scal/fun/grad_2F1.hpp

```diff
@@ -1,42 +1,109 @@
+#include <limits>
 
 namespace stan {
   namespace math {
 
-    // Gradient of the hypergeometric function 2F1(a, b | c | z)
-    // with respect to a and c
+    /**
+     * Gradients of the hypergeometric function, 2F1.
+     *
+     * Calculate the gradients of the hypergeometric function (2F1)
+     * as the power series stopping when the series converges
+     * to within <code>precision</code> or throwing when the
+     * function takes <code>max_steps</code> steps.
+     *
+     * This power-series representation converges for all gradients
+     * under the same conditions as the 2F1 function itself.
+     *
+     * @tparam T type of arguments and result
+     * @param[out] g_a1 g_a1 pointer to array of six values of type T, result.
+     * @param[out] g_b1 g_b1 pointer to array of six values of type T, result.
+     * @param[in] a1 a1 see generalized hypergeometric function definition.
+     * @param[in] a2 a2 see generalized hypergeometric function definition.
+     * @param[in] b1 b1 see generalized hypergeometric function definition.
+     * @param[in] z z see generalized hypergeometric function definition.
+     * @param[in] precision precision of the infinite sum. defaults to 1e-6
+     * @param[in] max_steps number of steps to take. defaults to 10,000
+     */
     template<typename T>
-    void grad_2F1(T& gradA, T& gradC, T a, T b, T c, T z, T precision = 1e-6) {
+    void grad_2F1(T& g_a1, T& g_b1, const T& a1, const T& a2, const T& b1,
+        const T& z, const T& precision = 1e-10, int max_steps = 1e5) {
+      check_2F1_converges("grad_2F1", a1, a2, b1, z);
+
+      using std::log;
       using std::fabs;
+      using std::exp;
+
+      g_a1 = 0.0;
+      g_b1 = 0.0;
+
+      T log_g_old[4];
+      for (T *q = log_g_old; q != log_g_old + 4; ++q)
+        *q = -1.0 * std::numeric_limits<double>::infinity();
+
+      T log_t_old = 0.0;
+      T log_t_new = 0.0;
 
-      gradA = 0;
-      gradC = 0;
+      T log_z = log(z);
 
-      T gradAold = 0;
-      T gradCold = 0;
+      T p = 0.0;
+
+      double log_t_new_sign = 1.0;
+      double log_t_old_sign = 1.0;
+      double log_g_old_sign[4];
+      for (double *q = log_g_old_sign; q != log_g_old_sign + 4; ++q)
+        *q = 1.0;
 
       int k = 0;
-      T tDak = 1.0 / (a - 1);
+      T term;
+      while (true) {
+        p = (a1 + k) * (a2 + k) / ((b1 + k) * (1 + k));
+        if (p == 0)
+          break;
+
+        log_t_new += log(fabs(p)) + log_z;
+        log_t_new_sign *= sign(p);
 
-      while (fabs(tDak * (a + (k - 1)) ) > precision || k == 0) {
-          const T r = ( (a + k) / (c + k) ) * ( (b + k) / (T)(k + 1) ) * z;
-          tDak = r * tDak * (a + (k - 1)) / (a + k);
+//      g_old[0] = t_new * (g_old[0] / t_old + 1.0 / (a1 + k));
+        term = log_g_old_sign[0] * log_t_old_sign *
+          exp(log_g_old[0] - log_t_old) + 1/(a1 + k);
+        log_g_old[0] = log_t_new + log(fabs(term));
+        log_g_old_sign[0] = sign(term) * log_t_new_sign;
```

Comment:
yeah, I'll replace the ones I can by the ternary operator w/ negation, but I won't change the ones that would require nested branching. Seem ok?